File: KaraokeShow/Internalization/EmbedResourceManager.vb
Imports System.ComponentModel
Imports System.Globalization
Imports System.IO
Imports System.Resources

Public Class EmbedResourceManager
    Inherits ComponentResourceManager

    Private _contextTypeInfo As Type
    Private _neutralResourcesCulture As CultureInfo


    Sub New(t As Type)
        MyBase.New(t)
        _contextTypeInfo = t
    End Sub

    Protected Overrides Function InternalGetResourceSet(culture As CultureInfo, createIfNotExists As Boolean, tryParents As Boolean) As ResourceSet
        Dim rs As ResourceSet = MyBase.InternalGetResourceSet(culture, createIfNotExists, tryParents)
        If rs Is Nothing Then
            Dim store As Stream = Nothing
            Dim resourceFilename As String = Nothing
            If Me._neutralResourcesCulture Is Nothing Then
                Me._neutralResourcesCulture = GetNeutralResourcesLanguage(Me.MainAssembly)
            End If
            If _neutralResourcesCulture.Equals(culture) Then culture = CultureInfo.InvariantCulture
            resourceFilename = GetResourceFileName(culture)
            store = Me.MainAssembly.GetManifestResourceStream(Me._contextTypeInfo, resourceFilename)
            If store IsNot Nothing Then
                rs = New ResourceSet(store)
            Else
                rs = MyBase.InternalGetResourceSet(culture, createIfNotExists, tryParents)
            End If
        End If
        Return rs
    End Function

End Class

File: KaraokeShow/Internationalization/EmbedResourceManager.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Resources;
using System.Text;

namespace MusicBeePlugin.Internationalization
{
    public class EmbedResourceManager : ComponentResourceManager
    {

        private Type _contextTypeInfo;
        private CultureInfo _neutralResourcesCulture;

        public EmbedResourceManager(Type t) : base(t)
        {
            _contextTypeInfo = t;
        }

        protected override ResourceSet InternalGetResourceSet(CultureInfo culture, bool createIfNotExists, bool tryParents)
        {
            var rs = (ResourceSet)ResourceSets[culture];
            if (rs == null)
            {
                Stream store = null;
                string resourceFilename = null;
                if (_neutralResourcesCulture == null)
                    _neutralResourcesCulture = GetNeutralResourcesLanguage(MainAssembly);
                if (_neutralResourcesCulture.Equals(culture))
                    culture = CultureInfo.InvariantCulture;
                resourceFilename = GetResourceFileName(culture);
                store = MainAssembly.GetManifestResourceStream(_contextTypeInfo, resourceFilename);
                if (store != null)
                {
                    rs = new ResourceSet(store);
                    AddResourceSet(ResourceSets, culture, ref rs);
                }
                else
                    rs = base.InternalGetResourceSet(culture, createIfNotExists, tryParents);
            }
            return rs;
        }

        private static void AddResourceSet(Hashtable localResourceSets, CultureInfo culture, ref ResourceSet rs)
        {
            lock (localResourceSets)
            {
                ResourceSet objA = (ResourceSet)localResourceSets[culture];
                if (objA != null)
                {
                    if (!object.Equals(objA, rs))
                    {
                        rs.Dispose();
                        rs = objA;
                    }
                }
                else
                {
                    localResourceSets.Add(culture, rs);
                }
            }
        }
    }
}
